### ISSUE TITLE
BUG in function project_crf_layer when using BERT + CRF without Bi-LSTM

### ISSUE THREAD

**Issue by passerbythesun** [macanv/BERT-BiLSTM-CRF-NER#254]
In function [`project_crf_layer`](https://github.com/macanv/BERT-BiLSTM-CRF-NER/blob/12fb822bd847f24e4de1beca8af9de34d1a794d2/bert_base/train/lstm_crf_layer.py#L147), activation function `tanh` is not needed. On the contrary, it leads to bad performance.

Many report unexpected results using this repo's code. I think this may be the problem.

Somebody check it!

**Comment by macanv** [comment]
THX， but, Why does the tanh function cause this problem?

**Comment by leonaxiongxin** [comment]
> THX， but, Why does the tanh function cause this problem?

as mentioned in  [issue 159](https://github.com/macanv/BERT-BiLSTM-CRF-NER/issues/159),

the 'tanh' function is wrong since it will map the logits to the interval [-1, 1], thus hurt the Viterbi decoding processing and the model performance.

the final output of `project_bilstm_layer` didn't have `tanh` too.

But I am wondering why there were two dense layers in function `project_bilstm_layer` , and `hidden = tf.tanh(tf.nn.xw_plus_b(output, W, b))` was  in  it?

**Comment by xmy7216** [comment]
> > THX， but, Why does the tanh function cause this problem?
> 
> as mentioned in [issue 159](https://github.com/macanv/BERT-BiLSTM-CRF-NER/issues/159),
> 
> the 'tanh' function is wrong since it will map the logits to the interval [-1, 1], thus hurt the Viterbi decoding processing and the model performance.
> 
> the final output of `project_bilstm_layer` didn't have `tanh` too.
> 
> But I am wondering why there were two dense layers in function `project_bilstm_layer` , and `hidden = tf.tanh(tf.nn.xw_plus_b(output, W, b))` was in it?

Hello，may I delete that line directly?

**Comment by leonaxiongxin** [comment]
> > > THX， but, Why does the tanh function cause this problem?
> > 
> > 
> > as mentioned in [issue 159](https://github.com/macanv/BERT-BiLSTM-CRF-NER/issues/159),
> > the 'tanh' function is wrong since it will map the logits to the interval [-1, 1], thus hurt the Viterbi decoding processing and the model performance.
> > the final output of `project_bilstm_layer` didn't have `tanh` too.
> > But I am wondering why there were two dense layers in function `project_bilstm_layer` , and `hidden = tf.tanh(tf.nn.xw_plus_b(output, W, b))` was in it?
> 
> Hello，may I delete that line directly?

use `tf.nn.xw_plus_b(output, W, b)` .

**Comment by xmy7216** [comment]
> > > > THX， but, Why does the tanh function cause this problem?
> > > 
> > > 
> > > as mentioned in [issue 159](https://github.com/macanv/BERT-BiLSTM-CRF-NER/issues/159),
> > > the 'tanh' function is wrong since it will map the logits to the interval [-1, 1], thus hurt the Viterbi decoding processing and the model performance.
> > > the final output of `project_bilstm_layer` didn't have `tanh` too.
> > > But I am wondering why there were two dense layers in function `project_bilstm_layer` , and `hidden = tf.tanh(tf.nn.xw_plus_b(output, W, b))` was in it?
> > 
> > 
> > Hello，may I delete that line directly?
> 
> use `tf.nn.xw_plus_b(output, W, b)` , without `tanh`

Thank you. I will try that way.

**Comment by macanv** [comment]
FIXED , some experience support that, remove tanh would imporove the stability and performance

**Comment by xmy7216** [comment]
> > THX， but, Why does the tanh function cause this problem?
> 
> as mentioned in [issue 159](https://github.com/macanv/BERT-BiLSTM-CRF-NER/issues/159),
> 
> the 'tanh' function is wrong since it will map the logits to the interval [-1, 1], thus hurt the Viterbi decoding processing and the model performance.
> 
> the final output of `project_bilstm_layer` didn't have `tanh` too.
> 
> But I am wondering why there were two dense layers in function `project_bilstm_layer` , and `hidden = tf.tanh(tf.nn.xw_plus_b(output, W, b))` was in it?

您好，我想请问一下，代码中两个全连接层有必要吗？是否需要去掉一个呢？

**Comment by Heaven-zhw** [comment]
> FIXED , some experience support that, remove tanh would imporove the stability and performance
Hello，the questioner probably means removing "tanh" the in the function "project_crf_layer", line 147, but you have removed it in the function "project_bilstm_layer" , line 119. Is that correct?

**Comment by Heaven-zhw** [comment]
> FIXED , some experience support that, remove tanh would imporove the stability and performance

Hello，the questioner probably means removing "tanh" the in the function "project_crf_layer", line 147, but you have removed it in the function "project_bilstm_layer" , line 119. Is that correct?

**Comment by leonaxiongxin** [comment]
> > FIXED , some experience support that, remove tanh would imporove the stability and performance
> 
> Hello，the questioner probably means removing "tanh" the in the function "project_crf_layer", line 147, but you have removed it in the function "project_bilstm_layer" , line 119. Is that correct?

Yeah, I removed `tanh` in `project_crf_layer` and it worker well. I mentioned `project_bilstm_layer` just to say the final output doesn't need `tanh`.

**Comment by Heaven-zhw** [comment]
> > > FIXED , some experience support that, remove tanh would imporove the stability and performance
> > 
> > 
> > Hello，the questioner probably means removing "tanh" the in the function "project_crf_layer", line 147, but you have removed it in the function "project_bilstm_layer" , line 119. Is that correct?
> 
> Yeah, I removed `tanh` in `project_crf_layer` and it worker well. I mentioned `project_bilstm_layer` just to say the final output doesn't need `tanh`.

Thanks，but it seems that the author did wrong in his last commit. You can check it: https://github.com/macanv/BERT-BiLSTM-CRF-NER/blob/master/bert_base/train/lstm_crf_layer.py , in the line 119 and 147. Is that code correct or not?
